src/main/java/teammates/ui/webapi/DeadlineExtensionsWorkerAction.java

```diff
@@ -28,129 +28,120 @@
+            emailsToSend.addAll(getDeadlineUpdatedEmails(course, session, deadlinesToUpdate, oldDeadlines, areInstructors));
         }
 
-        return emailGenerator.generateDeadlineExtensionEmail(course, session, oldEndTime, endTime,
-                EmailType.DEADLINE_EXTENSION_UPDATED, userEmail, isInstructor);
     }
 
-    private EmailWrapper revokeDeadlineExtension(CourseAttributes course, FeedbackSessionAttributes session,
-            String userEmail, boolean isInstructor, boolean notifyUsers) {
-        DeadlineExtensionAttributes deadlineExtension =
-                logic.getDeadlineExtension(course.getId(), session.getFeedbackSessionName(), userEmail, isInstructor);
-        if (deadlineExtension == null) {
-            log.severe("Unexpected error while revoking deadline extension");
-            return null;
-        }
-
-        logic.deleteDeadlineExtension(course.getId(), session.getFeedbackSessionName(), userEmail, isInstructor);
+    private List<EmailWrapper> getDeadlineRevokedEmails(CourseAttributes course,
+            FeedbackSessionAttributes session, Map<String, Instant> deadlines, boolean areInstructors) {
+        return deadlines.entrySet()
+                .stream()
+                .map(entry ->
+                        emailGenerator.generateDeadlineExtensionEmail(course, session, entry.getValue(),
+                                session.getEndTime(), EmailType.DEADLINE_EXTENSION_REVOKED, entry.getKey(), areInstructors))
+                .collect(Collectors.toList());
+    }
 
-        if (!notifyUsers) {
-            return null;
-        }
+    private List<EmailWrapper> getDeadlineCreatedEmails(CourseAttributes course,
+            FeedbackSessionAttributes session, Map<String, Instant> deadlines, boolean areInstructors) {
+        return deadlines.entrySet()
+                .stream()
+                .map(entry ->
+                        emailGenerator.generateDeadlineExtensionEmail(course, session, session.getEndTime(),
+                                entry.getValue(), EmailType.DEADLINE_EXTENSION_GIVEN, entry.getKey(), areInstructors))
+                .collect(Collectors.toList());
+    }
 
-        return emailGenerator.generateDeadlineExtensionEmail(course, session, deadlineExtension.getEndTime(),
-                session.getEndTime(), EmailType.DEADLINE_EXTENSION_REVOKED, userEmail, isInstructor);
+    private List<EmailWrapper> getDeadlineUpdatedEmails(CourseAttributes course, FeedbackSessionAttributes session,
+            Map<String, Instant> deadlines, Map<String, Instant> oldDeadlines, boolean areInstructors) {
+        return deadlines.entrySet()
+                .stream()
+                .map(entry ->
+                        emailGenerator.generateDeadlineExtensionEmail(course, session, oldDeadlines.get(entry.getKey()),
+                                entry.getValue(), EmailType.DEADLINE_EXTENSION_UPDATED, entry.getKey(), areInstructors))
+                .collect(Collectors.toList());
```

Comment:
Tiny nit: would it be better to have `getDeadlineOpsEmails` that takes in the `EmailType` as input so as to reduce code duplication?

Reply:
I can try, but the difference is not as simple as different EmailType. Notice the start and end times are different for each method. Also, getDeadlineUpdatedEmails takes in an extra parameter.

Reply:
Nvm! Can leave it!